### PR TITLE
(1) Download QPs for all known languages

### LINKED ISSUE #4275: [Bug] All QPs for supported languages are not being downloaded
### Description

The implementation of #4142 is not correct; it is only downloading a subset of the supported languages. It should attempt to download the QPs for all known languages.

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Fixes #4275

### CODE REVIEW

File: src/Integration/Binding/BindingProcessImpl.cs [hunk 1]
```diff
@@ -196,7 +216,7 @@ public bool FinishSolutionBindingOnUIThread()
         #region Private methods
 
         internal /* for testing */ IEnumerable<Language> GetBindingLanguages()
-            => host.SupportedPluginLanguages;
```
Comment (duncanp-sonar) [inline src/Integration/Binding/BindingProcessImpl.cs:199]: I changed this code to return this value earlier in the sprint (previously there was a lot of complicated code to walk the projects in the solution and find the languages that are actually used).

However, I now realise my original change is original. I should have changed it to return the list of all languages known to SLVS.

This change fixes that.

File: src/Integration/Binding/BindingProcessImpl.cs [hunk 2]
```diff
@@ -52,13 +53,32 @@ internal class BindingProcessImpl : IBindingProcess
...
+        { }
+
+        internal /* for testing */ BindingProcessImpl(IHost host,
+            BindCommandArgs bindingArgs,
+            ISolutionBindingOperation solutionBindingOperation,
+            IBindingConfigProvider bindingConfigProvider,
+            SonarLintMode bindingMode,
+            IExclusionSettingsStorage exclusionSettingsStorage,
+            bool isFirstBinding,
+            IEnumerable<Language> languagesToBind)
```
Comment (duncanp-sonar) [inline src/Integration/Binding/BindingProcessImpl.cs:73]: Added an internal constructor so we can inject the list of languages when testing.

File: src/Integration.UnitTests/Binding/BindingProcessImplTests.cs [hunk 3]
```diff
@@ -340,6 +332,19 @@ public void GetBindingLanguages_ReturnsExpectedLanguagese()
             CollectionAssert.AreEquivalent(expectedLanguages, actualLanguages.ToArray(), "Unexpected languages for binding projects");
         }
 
+        [TestMethod]
+        public void GetBindingLanguages_PublicCtor_ReturnsAllKnownLanguages()
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/Binding/BindingProcessImplTests.cs:336]: New test that the default behaviour returns all known languages.

File: src/Integration.UnitTests/Binding/BindingProcessImplTests.cs [hunk 4]
```diff
@@ -325,13 +318,12 @@ public async Task DownloadQualityProfile_WhenBindingConfigIsNull_Fails()
         }
 
         [TestMethod]
-        public void GetBindingLanguages_ReturnsExpectedLanguagese()
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/Binding/BindingProcessImplTests.cs:328]: Miscellaneous: fixed typo _...Languagese_

File: src/Integration.UnitTests/Binding/BindingProcessImplTests.cs [hunk 5]
```diff
@@ -188,9 +188,7 @@ public async Task DownloadQualityProfile_Success()
                 .ReturnsAsync(bindingConfig);
 
             var bindingArgs = CreateBindCommandArgs("key", ProjectName, new ConnectionInformation(new Uri("http://connected")));
-            var testSubject = this.CreateTestSubject(bindingArgs, configProviderMock.Object);
-
-            host.SupportedPluginLanguages.Add(language);
+            var testSubject = this.CreateTestSubject(bindingArgs, configProviderMock.Object, languagesToBind: new[] { language });
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/Binding/BindingProcessImplTests.cs:191]: Fixed up existing tests (most of the tests only expected a few binding languages).

**Review by bigfluffycookie — APPROVED** [review]
Nice!

**Pull request merged**